lib/Sema/CSSolver.cpp

```diff
@@ -2430,39 +2427,200 @@ Constraint *getApplicableFunctionConstraint(ConstraintSystem &CS,
   return found;
 }
 
-static unsigned countUnboundTypes(ConstraintSystem &CS,
-                                  Constraint *disjunction) {
-  auto *nested = disjunction->getNestedConstraints().front();
-  assert(nested->getKind() == ConstraintKind::BindOverload &&
-         "Expected bind overload constraint!");
+static float scoreTypeVariable(ConstraintSystem &cs,
+                               TypeVariableType *typeVar) {
+  if (auto fixedType = cs.getFixedType(typeVar))
+    return 1.0;
+
+  // If there is no associated type for this type variable
+  // let's see if it has any conversion constraints, which
+  // also contribute to search pruning.
+
+  SmallVector<Constraint *, 4> constraints;
+  cs.getConstraintGraph().gatherConstraints(
+      typeVar, constraints, ConstraintGraph::GatheringKind::EquivalenceClass);
+
+  for (auto constraint : constraints) {
+    if (constraint->getKind() != ConstraintKind::Conversion)
```

Comment:
There are other constraints like SubType constraints that have similar properties. I suggest switching on `constraint->getKind()` so it's clear that we've thought about all of these.